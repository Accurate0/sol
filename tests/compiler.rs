use insta::assert_debug_snapshot;
use plrs::{compiler::Compiler, lexer::Lexer, parser::Parser};

#[test]
fn small_input() {
    let input = r#"
let x = 3;
let y = 4;
let z = x + y;

fn print() {}

fn test(a) {
    let y = 1.3 + a;
    {
        let z = y + 3;
    }

    let z = y + 2;
}

fn main() {
    let x = 1.3 + 3;
    {
	print("Hello");
	print(x);
	let y = test(4);
    }

    print(x);
}


main();
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let output = compiler.compile().unwrap();

    assert_debug_snapshot!(output);
}

#[test]
fn variable_mutation() {
    let input = r#"
let mut x = 3;
x = 2;
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let output = compiler.compile().unwrap();

    assert_debug_snapshot!(output);
}

#[test]
fn prefix() {
    let input = r#"
let x = -3;
let y = -(x + 3);
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let output = compiler.compile().unwrap();

    assert_debug_snapshot!(output);
}

#[test]
fn prefix_boolean() {
    let input = r#"
let x = true;
let y = !x;
        "#
    .to_owned();

    let mut lexer = Lexer::new(&input);
    let mut parser = Parser::new(&mut lexer, &input);
    let compiler = Compiler::new(&mut parser);

    let output = compiler.compile().unwrap();

    assert_debug_snapshot!(output);
}
