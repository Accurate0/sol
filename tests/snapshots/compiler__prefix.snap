---
source: tests/compiler.rs
expression: output
---
CompiledProgram {
    functions: [],
    global_code: [
        LoadLiteral {
            dest: 1,
            src: 0,
        },
        PrefixSub {
            dest: 2,
            rhs: 1,
        },
        LoadLiteral {
            dest: 3,
            src: 0,
        },
        Add {
            dest: 4,
            lhs: 2,
            rhs: 3,
        },
        PrefixSub {
            dest: 5,
            rhs: 4,
        },
    ],
    global_register_count: 6,
    literals: [
        Integer(
            3,
        ),
    ],
}
